cranelift/codegen/src/loop_analysis.rs

```diff
@@ -229,16 +230,18 @@ impl LoopAnalysis {
         // We handle each loop header in reverse order, corresponding to a pseudo postorder
         // traversal of the graph.
         for lp in self.loops().rev() {
-            for BlockPredecessor {
-                block: pred,
-                inst: pred_inst,
-            } in cfg.pred_iter(self.loops[lp].header)
-            {
-                // We follow the back edges
-                if domtree.dominates(self.loops[lp].header, pred_inst, layout) {
-                    stack.push(pred);
-                }
-            }
+            stack.extend(
+                cfg.pred_iter(self.loops[lp].header)
+                    .filter(
+                        |BlockPredecessor {
+                             inst: pred_inst, ..
+                         }| {
+                            // We follow the back edges
+                            domtree.dominates(self.loops[lp].header, *pred_inst, layout)
+                        },
+                    )
+                    .map(|BlockPredecessor { block, .. }| block),
+            );
```

Comment:
I'm not sure how I feel about this one either -- is this really more readable than the for-loop with an if-statement?

Reply:
I think some of the complexity is in how I wrote the de structuring - changing it to `pred.inst`, as opposed to doing it between the pipes. Also would a collect be most appropriate here - used extend to ensure exact same semantics as before with push, but the `while let should` drain  `stack` and so it is being extended from empty each time (in which case collect may make this more clear)